{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "hw1_110761504.ipynb",
      "provenance": [],
      "authorship_tag": "ABX9TyOa25N+e+MduUm80lpPmfeX",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mequanent/Bioinformatics-hws/blob/main/hws_110761504.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Clone the repository first using the following command\n",
        "!npx degit https://github.com/mequanent/Bioinformatics-hw2 -f"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Eim-tkyVaZrg",
        "outputId": "248d9204-9bab-4e85-8a3c-d56c1a1aafb7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[K\u001b[?25hnpx: installed 1 in 2.22s\n",
            "\u001b[36m> destination directory is not empty. Using --force, continuing\u001b[39m\n",
            "\u001b[36m> cloned \u001b[1mmequanent/Bioinformatics-hw2\u001b[22m#\u001b[1mHEAD\u001b[22m\u001b[39m\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Homework 2**"
      ],
      "metadata": {
        "id": "GCOZucxBv9DA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "pip install Bio -q "
      ],
      "metadata": {
        "id": "Tba1tOZ5vhIb",
        "outputId": "f9e44ea5-edb2-4c87-e1ae-e83db48265fe",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[K     |████████████████████████████████| 269 kB 5.2 MB/s \n",
            "\u001b[K     |████████████████████████████████| 2.3 MB 46.7 MB/s \n",
            "\u001b[?25h"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "#To install the Bio library into my IDLE python editor I used the following two steps.\n",
        "#    1. cd C:\\Users\\USER\\AppData\\Local\\Programs\\Python\\Python39\\Scripts\n",
        "#    2. pip install Bio\n",
        "\n",
        "# Read fasta and parse the seq_names and the sequences\n",
        "# Reference for parsing the fasta is the original biopython library documentation\n",
        "# (https://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc11 section 2.4.1)\n",
        "\n",
        "from Bio import SeqIO\n",
        "seq_name = []\n",
        "sequence = []\n",
        "# Remember to customize as --input from CLI args\n",
        "for seq_record in SeqIO.parse(\"test.fasta\", \"fasta\"):\n",
        "  seq_name.append(seq_record.id)\n",
        "  #sequence.append(repr(seq_record.seq))\n",
        "  sequence.append(seq_record.seq)\n",
        "  #print(len(seq_record))\n",
        "print(sequence[0])\n",
        "print(sequence[1])\n",
        "\n",
        "def preprocess_score(score):\n",
        "  file = open(score, 'r')\n",
        "  lines = file.readlines()\n",
        "  file.close()\n",
        "  columns = []\n",
        "  mut = []\n",
        "  for line in lines:\n",
        "    if line.startswith(\"#\") or line.startswith(\" \"):\n",
        "      continue\n",
        "    columns.append(line[0]) # Amino acid alphabets stored for row & column indexing\n",
        "    line = line[1:].strip().split() # remove the letters from left and the spaces\n",
        "    mut.append(line)\n",
        "  mut = np.array(mut).astype(int) # change the string numbers into type int\n",
        "  \n",
        "  return pd.DataFrame(mut, index = columns, columns = columns)\n",
        "\n",
        "score = \"pam250.txt\" # score from command line args\n",
        "score = preprocess_score(score) # score made a data frame to access its values with \n",
        "                  # alphabet indices\n",
        "#print(score)\n",
        "\n",
        "gop = -10\n",
        "gex = -2\n",
        "# A function to determine pairwise alignment score\n",
        "def pairwise(x, y):\n",
        "  pair_sum = 0\n",
        "  x_prev = '' # x_prev and y_prev are used to store immediate previous alphabets\n",
        "  y_prev = '' # so that extension gaps will be identified\n",
        "  for i in range(len(x)): # Considering that both x and y have the same length\n",
        "    x_prev = x_prev == '-' and x[i] == '-' # status of extension gap in the first paired sequence\n",
        "    y_prev = y_prev == '-' and y[i] == '-' # status of extension gap in second paired sequence\n",
        "    curr = x[i] == '-' or y[i] == '-' # status of open gap in the paired sequence\n",
        "    if x_prev or y_prev: # extension gap value will be taken if holds\n",
        "      pair_sum = pair_sum + gex\n",
        "    elif curr: # open gap value will be taken if holds\n",
        "      pair_sum = pair_sum + gop\n",
        "    else: # score matrix value will be taken otherwise\n",
        "      pair_sum = pair_sum + score.loc[x[i]][y[i]] \n",
        "    x_prev = x[i]\n",
        "    y_prev = y[i]\n",
        "  return pair_sum\n",
        "\n",
        "#pairwise(sequence[0], sequence[1])\n",
        "\n",
        "# A function to find the total sum of pairs score.\n",
        "def sumOfPairs(seq):\n",
        "  sop = 0 # to store sum of pairs value\n",
        "  for i in range(len(seq)-1): # i iterates as long as there is j in seq as its next sequence\n",
        "    for j in range(i+1, len(seq)): # j will be the next seq index to be compared with seq[i]\n",
        "      sop = sop + pairwise(seq[i], seq[j])\n",
        "  return sop\n",
        "print(sumOfPairs(sequence))"
      ],
      "metadata": {
        "id": "UlUbMGY6cAcx",
        "outputId": "5a47f0a6-78ab-48dc-bc11-8b0fe7756569",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "N-LFVALYDFVASGDNTLSITKGEKLRVLGYNH--NGEWCEAQTKNGQGWVPSNYITPVN\n",
            "KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP\n",
            "1047\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Homework 3**"
      ],
      "metadata": {
        "id": "whhnJvDTvyh4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "seq_name = []\n",
        "sequence = []\n",
        "# Remember to customize as --input from CLI args\n",
        "for seq_record in SeqIO.parse(\"test3.fasta\", \"fasta\"):\n",
        "  seq_name.append(seq_record.id)\n",
        "  sequence.append(seq_record.seq)\n",
        "\n",
        "seq1 = sequence[0]\n",
        "seq2 = sequence[1]\n",
        "\n",
        "print(\"+----- Given Sequences with their respective lengthes ---------+\")\n",
        "print(seq1, len(seq1))\n",
        "print(seq2, len(seq2))\n",
        "print(\"****************************************************************\")\n",
        "gap = -10\n",
        "score = \"pam250.txt\" # score from command line args\n",
        "aln = 'local' # Alignment choice argument from command line interface\n",
        "\n",
        "TOP = gap\n",
        "DIAGONAL = 0\n",
        "LEFT = gap\n",
        "col = '-' + seq1 # This if for simplicity to avoid repeated use of (len(seq1) + 1)\n",
        "row = '-' + seq2 # For the same reason\n",
        "\n",
        "def preprocess_score(score):\n",
        "  file = open(score, 'r')\n",
        "  lines = file.readlines()\n",
        "  file.close()\n",
        "  columns = []\n",
        "  mut = []\n",
        "  for line in lines:\n",
        "    if line.startswith(\"#\") or line.startswith(\" \"):\n",
        "      continue\n",
        "    columns.append(line[0]) # Amino acid alphabets stored for row & column indexing\n",
        "    line = line[1:].strip().split() # remove the letters from left and the spaces\n",
        "    mut.append(line)\n",
        "  mut = np.array(mut).astype(int) # change the string numbers into type int\n",
        "  return pd.DataFrame(mut, index = columns, columns = columns)\n",
        "\n",
        "score = preprocess_score(score) # score made a data frame to access its values  \n",
        "                  # with alphabet indices\n",
        "\n",
        "# *************************** Global Alignment Section *************************\n",
        "# Filling score_m and keeping track of the direction for each cell\n",
        "def globalAlignment(score):\n",
        "  score_m = np.zeros((len(row), len(col))) # score matrix for global alignment \n",
        "  direction_m = np.empty((len(row), len(col)), dtype = object)\n",
        "  for i in range(len(col)):\n",
        "    score_m[0][i] = i * gap\n",
        "  for j in range(len(row)):\n",
        "    score_m[j][0] = j * gap\n",
        "\n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      TOP = score_m[i-1,j] + gap\n",
        "      DIAGONAL = score_m[i-1][j-1] + score.loc[row[i]][col[j]]\n",
        "      LEFT = score_m[i,j-1] + gap\n",
        "      score_m[i][j] = max(TOP, DIAGONAL, LEFT) \n",
        "      # ------ Let the direction for each cell be saved as follows \n",
        "      if score_m[i][j] == TOP:\n",
        "        direction_m[i][j] = 'ins'\n",
        "      elif score_m[i][j] == DIAGONAL:\n",
        "        direction_m[i][j] = 'sub' \n",
        "      else:\n",
        "        direction_m[i][j] = 'del'\n",
        "\n",
        "  score_m = score_m.astype(int)\n",
        "  return score_m, direction_m\n",
        "\n",
        "# Trace back section for global alignment\n",
        "def traceBackGlobal():\n",
        "  score_m, direction_m = globalAlignment(score)\n",
        "  i = len(seq1)\n",
        "  j = len(seq2)\n",
        "  aln1 = ''\n",
        "  aln2 = ''\n",
        "\n",
        "  while(not (i == 0 and j == 0)):\n",
        "    if(direction_m[j][i] == 'sub'):\n",
        "      aln1 = col[i] + aln1\n",
        "      aln2 = row[j] + aln2\n",
        "      i, j = i - 1, j - 1\n",
        "    elif(direction_m[j][i] == 'del'):\n",
        "      aln1 = col[i] + aln1\n",
        "      aln2 = '-' + aln2\n",
        "      i = i - 1\n",
        "    else:\n",
        "      aln1 = '-' + aln1\n",
        "      aln2 = row[j] + aln2\n",
        "      j = j - 1\n",
        "  return aln1, aln2\n",
        "# End of Trace Back for the Global Alignment\n",
        "# --------------------------------------------\n",
        "# ************* End of Global Alignment ************************************\n",
        "\n",
        "def showMatched(x, y): # A function to help proper display of the alignments\n",
        "  between = ''\n",
        "  for i in range(len(x)):\n",
        "    if x[i] == y[i]:\n",
        "      between = between + '|'\n",
        "    elif x[i] == '-' or y[i] == '-':\n",
        "      between = between + ' '\n",
        "    else:\n",
        "      between = between + '.'\n",
        "  return between\n",
        "\n",
        "def displayAlignment(x, y): # A function to display the aligned sequences with their SoP score\n",
        "  print(\"+----- You wanted {} alignment and here is the result -----+\".format(aln))\n",
        "  print(x)\n",
        "  print(showMatched(x, y))\n",
        "  print(y)\n",
        "  print(\"Sum of pairs (SoP) score for the {a} alignment is: {b}\".format(a = aln, b = pairwise(x, y)))\n",
        "# The SoP value verifies the correctness of our alignment\n",
        "                # if it becomes equal to the right bottom value.\n",
        "\n",
        "# ************* Local Alignment ********************************************\n",
        "def localAlignment(score):\n",
        "  score_local = np.zeros((len(row), len(col))).astype(int) # score matrix for local alignment\n",
        "  direction_l = np.zeros((len(row), len(col)), dtype = object) # to trace direction\n",
        "  bestScore = 0\n",
        "  \n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      TOP = score_local[i-1,j] + gap\n",
        "      DIAGONAL = score_local[i-1][j-1] + score.loc[row[i]][col[j]]\n",
        "      LEFT = score_local[i,j-1] + gap\n",
        "      score_local[i][j] = max(TOP, DIAGONAL, LEFT, 0)\n",
        "\n",
        "      if score_local[i][j] >= bestScore:\n",
        "        bestScore = score_local[i][j]\n",
        "        \n",
        "      if score_local[i][j] == DIAGONAL:\n",
        "        direction_l[i][j] = 'sub'\n",
        "      elif score_local[i][j] == TOP:\n",
        "        direction_l[i][j] = 'ins' \n",
        "      else:\n",
        "        direction_l[i][j] = 'del'\n",
        "  return score_local, direction_l, bestScore\n",
        "\n",
        "# Trace back all possible optimal local alignment\n",
        "def traceBackLocal():\n",
        "  score_local, direction_l, bestScore = localAlignment(score)\n",
        "\n",
        "  r, c = score_local.shape # row and col lengths to track best score locations\n",
        "  bestLoc = [] # list of best score cells\n",
        "  pairs = [] # list of local pairs\n",
        "  \n",
        "  # Find all locations of the best score from the local alignment score matrix\n",
        "  for i in range(1, r):\n",
        "    for j in range(1, c):\n",
        "      if score_local[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "  \n",
        "  for k in range(len(bestLoc)): # loop over all optimal local alignments\n",
        "    i, j = bestLoc[k]\n",
        "    al1 = ''\n",
        "    al2 = ''\n",
        "    between = ''\n",
        "    while(score_local[i][j]> 0):\n",
        "      if(direction_l[i][j] == 'sub'):\n",
        "        al1 = col[j] + al1\n",
        "        al2 = row[i] + al2\n",
        "        i, j = i - 1, j - 1\n",
        "      elif(direction_l[i][j] == 'del'):\n",
        "        al1 = col[j] + al1\n",
        "        al2 = '-' + al2\n",
        "        j = j - 1\n",
        "      else:\n",
        "        al1 = '-' + al1\n",
        "        al2 = row[i] + al2\n",
        "        i = i - 1\n",
        "    pairs.append((al1, al2))\n",
        "  return pairs\n",
        "\n",
        "alignments = []\n",
        "if aln == 'global':\n",
        "  aln1, aln2 = traceBackGlobal()\n",
        "  displayAlignment(aln1, aln2)\n",
        "  alignments.append(SeqRecord(Seq(aln1), id=\"protien 1\"))\n",
        "  alignments.append(SeqRecord(Seq(aln2), id=\"protien 2\"))\n",
        "elif aln == 'local':\n",
        "  for i in traceBackLocal():\n",
        "    displayAlignment(i[0], i[1])\n",
        "    alignments.append(SeqRecord(Seq(i[0]), id=\"protien 1\"))\n",
        "    alignments.append(SeqRecord(Seq(i[1]), id=\"protien 2\"))\n",
        "else:\n",
        "  print(\"Please selected the alignment type.\")\n",
        "\n",
        "SeqIO.write(alignments, \"result.fasta\", \"fasta\")"
      ],
      "metadata": {
        "id": "e_YVQGnpvxOa",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4bdf4c48-182b-4e9f-9ef0-152f5d1f3ac6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "+----- Given Sequences with their respective lengthes ---------+\n",
            "NLFVALYDFVASGDNTLSITKGEKLRVLGYNHNGEWCEAQTKNGQGWVPSNYITPVN 57\n",
            "KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP 60\n",
            "****************************************************************\n",
            "+----- You wanted local alignment and here is the result -----+\n",
            "LFVALYDFVASGDNTLSITKGEKLRVLG-YNHNG-EWCEAQTKNGQGWVPSNYI\n",
            "...||.|.....|..|....|....... ..... ||..|......|.||.|..\n",
            "VIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLL\n",
            "Sum of pairs (SoP) score for the local alignment is: 52\n",
            "+----- You wanted local alignment and here is the result -----+\n",
            "LFVALYDFVASGDNTLSITKGEKLRVLG-YNHNG-EWCEAQTKNGQGWVPSNYIT\n",
            "...||.|.....|..|....|....... ..... ||..|......|.||.|...\n",
            "VIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLG\n",
            "Sum of pairs (SoP) score for the local alignment is: 52\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "4"
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "***HW4:***\n",
        " Global and Local Alignments with Affine Gap Penalties"
      ],
      "metadata": {
        "id": "FqM152ET55WU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "seq_name = []\n",
        "sequence = []\n",
        "\n",
        "for seq_record in SeqIO.parse(\"test3.fasta\", \"fasta\"):\n",
        "  seq_name.append(seq_record.id)\n",
        "  sequence.append(seq_record.seq)\n",
        "\n",
        "seq1 = sequence[1]\n",
        "seq2 = sequence[0]\n",
        "\n",
        "print(\"+----- Given Sequences with their respective lengthes ---------+\")\n",
        "print(seq1, len(seq1))\n",
        "print(seq2, len(seq2))\n",
        "print(\"****************************************************************\")\n",
        "gop= -10\n",
        "gex = -2\n",
        "score = \"pam250.txt\" # score from command line args\n",
        "aln = 'local' # Alignment choice argument from command line interface\n",
        "\n",
        "col = '-' + seq1 # This if for simplicity to avoid repeated use of (len(seq1) + 1)\n",
        "row = '-' + seq2 # For the same reason\n",
        "\n",
        "def preprocess_score(score):\n",
        "  file = open(score, 'r')\n",
        "  lines = file.readlines()\n",
        "  file.close()\n",
        "  columns = []\n",
        "  mut = []\n",
        "  for line in lines:\n",
        "    if line.startswith(\"#\") or line.startswith(\" \"):\n",
        "      continue\n",
        "    columns.append(line[0]) # Amino acid alphabets stored for row & column indexing\n",
        "    line = line[1:].strip().split() # remove the letters from left and the spaces\n",
        "    mut.append(line)\n",
        "  mut = np.array(mut).astype(int) # change the string numbers into type int\n",
        "  return pd.DataFrame(mut, index = columns, columns = columns)\n",
        "\n",
        "score = preprocess_score(score) # score made a data frame to access its values  \n",
        "                  # with alphabet indices\n",
        "\n",
        "def createMatrices():\n",
        "  M = np.zeros((len(row), len(col))) # score matrix for global alignment \n",
        "  Ix = np.zeros((len(row), len(col)))\n",
        "  Iy = np.zeros((len(row), len(col)))\n",
        "\n",
        "  traceM = np.zeros((len(row), len(col)), dtype = object)\n",
        "  traceIx = np.zeros((len(row), len(col)), dtype = object)\n",
        "  traceIy = np.zeros((len(row), len(col)), dtype = object)\n",
        "\n",
        "  return M, Ix, Iy, traceM, traceIx, traceIy\n",
        "\n",
        "# ***************** Global Alignment Section *************************\n",
        "def globalAlignment():\n",
        "  # Filling score matrices and keeping track of the direction for each cell\n",
        "  M, Ix, Iy, traceM, traceIx, traceIy = createMatrices()\n",
        "\n",
        "  # Initializing the first column and first row\n",
        "  Ix[0][0] = Iy[0][0] = gop\n",
        "  for i in range(1, len(row)):\n",
        "    M[i][0] = Iy[i][0] = np.NINF\n",
        "    Ix[i][0] = gop + i * gex \n",
        "  for j in range(1, len(col)):\n",
        "    M[0][j] = Ix[0][j] = np.NINF\n",
        "    Iy[0][j] = gop + j * gex\n",
        "\n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      mat = score.loc[row[i]][col[j]]\n",
        "      M[i][j] = max((M[i-1][j-1] + mat), (Ix[i-1][j-1] + mat), (Iy[i-1][j-1] + mat)) \n",
        "      Ix[i,j] = max((M[i - 1, j] + gop), (Ix[i - 1, j] + gex))\n",
        "      Iy[i,j] = max((M[i, j - 1] + gop), (Iy[i, j - 1] + gex))\n",
        "\n",
        "      if M[i][j] == (Ix[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'x'\n",
        "      elif M[i][j] == (Iy[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'y'\n",
        "      else:\n",
        "        traceM[i][j] = 'm'\n",
        "\n",
        "      if Ix[i][j] == (Ix[i - 1, j] + gex):\n",
        "        traceIx[i][j] = 'x'\n",
        "      else:\n",
        "        traceIx[i][j] = 'm'\n",
        "      \n",
        "      if Iy[i][j] == (Iy[i, j - 1] + gex):\n",
        "        traceIy[i][j] = 'y'\n",
        "      else:\n",
        "        traceIy[i][j] = 'm'   \n",
        "\n",
        "# Trace back section for global alignment\n",
        "  j = len(seq1)\n",
        "  i = len(seq2)\n",
        "  aln1 = ''\n",
        "  aln2 = ''\n",
        "  mxm = max(M[i][j], Ix[i][j], Iy[i][j])\n",
        "  tracer = ''\n",
        "  if mxm == M[i][j]:\n",
        "    tracer = 'm'\n",
        "  elif mxm == Ix[i][j]:\n",
        "    tracer = 'x'\n",
        "  else:\n",
        "    tracer = 'y'\n",
        "    \n",
        "  while (i > 0 and j > 0):\n",
        "    if tracer == 'm':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      if traceM[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceM[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      elif traceM[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      i, j = i - 1, j - 1\n",
        "      continue\n",
        "    \n",
        "    elif tracer == 'x':\n",
        "      aln1 = '-' + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      if traceIx[i][j] == 'm':\n",
        "        tracer = 'm'    \n",
        "      elif traceIx[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      i = i - 1\n",
        "      continue\n",
        "      \n",
        "    elif tracer == 'y':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = '-' + aln2\n",
        "      if traceIy[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceIy[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      j = j - 1\n",
        "    else:\n",
        "      break\n",
        "      \n",
        "  return aln1, aln2\n",
        "# --------------------------------------------\n",
        "# ************* End of Global Alignment ************************************\n",
        "\n",
        "def isbest(a, b):\n",
        "  if float(a) > float(b):\n",
        "    return a\n",
        "  else:\n",
        "    return b\n",
        "# -------------------------------------------------------------------\n",
        "# Local Alignment Section \n",
        "def localAlignment():\n",
        "  bestScore = 0.0\n",
        "  # Filling score matrices and keeping track of the direction for each cell\n",
        "  M, Ix, Iy, traceM, traceIx, traceIy = createMatrices()\n",
        "\n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      mat = score.loc[row[i]][col[j]]\n",
        "      M[i][j] = max((M[i-1][j-1] + mat), (Ix[i-1][j-1] + mat), (Iy[i-1][j-1] + mat), 0.0) \n",
        "      Ix[i,j] = max((M[i - 1, j] + gop), (Ix[i - 1, j] + gex), 0.0)\n",
        "      Iy[i,j] = max((M[i, j - 1] + gop), (Iy[i, j - 1] + gex), 0.0)\n",
        "\n",
        "      if M[i][j] == (Ix[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'x'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "      elif M[i][j] == (Iy[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'y'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "      elif M[i][j] == (M[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'm'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "\n",
        "      if Ix[i][j] == (Ix[i - 1, j] + gex):\n",
        "        traceIx[i][j] = 'x'\n",
        "        bestScore = isbest(Ix[i][j], bestScore)\n",
        "      elif Ix[i][j] == (M[i - 1, j] + gop):\n",
        "        traceIx[i][j] = 'm'\n",
        "        bestScore = isbest(Ix[i][j], bestScore)\n",
        "      \n",
        "      if Iy[i][j] == (Iy[i, j - 1] + gex):\n",
        "        traceIy[i][j] = 'y'\n",
        "        bestScore = isbest(Iy[i][j], bestScore)\n",
        "      elif Iy[i][j] == (M[i, j - 1] + gop):\n",
        "        traceIy[i][j] = 'm'   \n",
        "        bestScore = isbest(Iy[i][j], bestScore)\n",
        "\n",
        "# Trace back section for local alignment\n",
        "  r, c = M.shape # row and col lengths to track best score locations\n",
        "  bestLoc = [] # list of best score cells\n",
        "  pairs = [] # list of local pairs\n",
        "  \n",
        "  # Find all locations of the best score from the local alignment score matrix\n",
        "  for i in range(1, r):\n",
        "    for j in range(1, c):\n",
        "      if M[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "      if Ix[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "      if Iy[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "  \n",
        "  for k in range(len(bestLoc)): # loop over all optimal local alignments\n",
        "    i, j = bestLoc[k]\n",
        "    aln1 = ''\n",
        "    aln2 = ''\n",
        "    mxm = max(M[i][j], Ix[i][j], Iy[i][j], 0)\n",
        "    tracer = ''\n",
        "    if mxm == M[i][j]:\n",
        "      tracer = 'm'\n",
        "    elif mxm == Ix[i][j]:\n",
        "      tracer = 'x'\n",
        "    else:\n",
        "      tracer = 'y'\n",
        "      \n",
        "    while (i > 0 and j > 0):# and tracer != ''):\n",
        "      if tracer == 'm' and M[i][j]>0:\n",
        "        aln1 = col[j] + aln1\n",
        "        aln2 = row[i] + aln2\n",
        "        if traceM[i][j] == 'm':\n",
        "          tracer = 'm'\n",
        "        elif traceM[i][j] == 'x':\n",
        "          tracer = 'x'\n",
        "        elif traceM[i][j] == 'y':\n",
        "          tracer = 'y'\n",
        "        i, j = i - 1, j - 1\n",
        "        continue\n",
        "      \n",
        "      elif tracer == 'x' and Ix[i][j] > 0:\n",
        "        aln1 = '-' + aln1\n",
        "        aln2 = row[i] + aln2\n",
        "        if traceIx[i][j] == 'm':\n",
        "          tracer = 'm'    \n",
        "        elif traceIx[i][j] == 'x':\n",
        "          tracer = 'x'\n",
        "        i = i - 1\n",
        "        continue\n",
        "        \n",
        "      elif tracer == 'y' and Iy[i][j] > 0:\n",
        "        aln1 = col[j] + aln1\n",
        "        aln2 = '-' + aln2\n",
        "        if traceIy[i][j] == 'm':\n",
        "          tracer = 'm'\n",
        "        elif traceIy[i][j] == 'y':\n",
        "          tracer = 'y'\n",
        "        j = j - 1\n",
        "      else:\n",
        "        break\n",
        "    pairs.append((aln1, aln2))\n",
        "  return pairs  \n",
        "# End of Local Alignment\n",
        "# --------------------------------------------\n",
        "\n",
        "def showMatched(x, y): # A function to help proper display of the alignments\n",
        "  between = ''\n",
        "  for i in range(len(x)):\n",
        "    if x[i] == y[i]:\n",
        "      between = between + '|'\n",
        "    elif x[i] == '-' or y[i] == '-':\n",
        "      between = between + ' '\n",
        "    else:\n",
        "      between = between + '.'\n",
        "  return between\n",
        "\n",
        "def displayAlignment(x, y): # A function to display the aligned sequences with their SoP score\n",
        "  print(x)\n",
        "  print(showMatched(x, y))\n",
        "  print(y)\n",
        "  print(\"Sum of pairs (SoP) score for the {a} alignment is: {b}\".format(a = aln, b = pairwise(x, y)))\n",
        "\n",
        "# The SoP value verifies the correctness of our alignment if it becomes\n",
        "# equal to the right bottom value for global and to the best score for local.\n",
        "\n",
        "# --------------------------------------------\n",
        "# **** Do the selected alignment by calling corresponding functions ***********\n",
        "print(\"+----- You wanted {} alignment and here is the result -----+\".format(aln))  \n",
        "alignments = []\n",
        "if aln == 'global':\n",
        "  aln1, aln2 = globalAlignment()\n",
        "  displayAlignment(aln1, aln2)\n",
        "  alignments.append(SeqRecord(Seq(aln1), id=\"protien 1\"))\n",
        "  alignments.append(SeqRecord(Seq(aln2), id=\"protien 2\"))\n",
        "elif aln == 'local':\n",
        "  j = 0\n",
        "  for i in localAlignment():\n",
        "    print(\"\\n****** Local Alignment: {} ******** \".format(j))\n",
        "    displayAlignment(i[0], i[1])\n",
        "    alignments.append(SeqRecord(Seq(i[0]), id=seq_name[0], description=\":|protien 1\"))\n",
        "    alignments.append(SeqRecord(Seq(i[1]), id=seq_name[1], description=\":|protien 2\"))\n",
        "    j += 1\n",
        "else:\n",
        "  print(\"Please select the alignment type.\")\n",
        "\n",
        "SeqIO.write(alignments, \"result.fasta\", \"fasta\")"
      ],
      "metadata": {
        "id": "TxfYaYuqUvx7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bc6110e3-292e-4ab6-86d1-7f82d54b72cc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "+----- Given Sequences with their respective lengthes ---------+\n",
            "KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP 60\n",
            "NLFVALYDFVASGDNTLSITKGEKLRVLGYNHNGEWCEAQTKNGQGWVPSNYITPVN 57\n",
            "****************************************************************\n",
            "+----- You wanted local alignment and here is the result -----+\n",
            "\n",
            "****** Local Alignment: 0 ******** \n",
            "VIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLL\n",
            "...||.|.....|..|....|............|  |..|......|.||.|..\n",
            "LFVALYDFVASGDNTLSITKGEKLRVLGYNHNGE--WCEAQTKNGQGWVPSNYI\n",
            "Sum of pairs (SoP) score for the local alignment is: 59\n",
            "\n",
            "****** Local Alignment: 1 ******** \n",
            "VIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLG\n",
            "...||.|.....|..|....|............|  |..|......|.||.|...\n",
            "LFVALYDFVASGDNTLSITKGEKLRVLGYNHNGE--WCEAQTKNGQGWVPSNYIT\n",
            "Sum of pairs (SoP) score for the local alignment is: 59\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "4"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Expected Output \n",
        "c = \"N-LFVALYDFVASGDNTLSITKGEKLRVLGYNHNGE--WCEAQTKNGQGWVPSNYITPVN\"\n",
        "d = \"KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP\"\n",
        "print(pairwise(c,d))\n",
        "# Another possible output with the same SoP score\n",
        "c = \"N-LFVALYDFVASGDNTLSITKGEKLRVLGYNHNG--EWCEAQTKNGQGWVPSNYITPVN\"\n",
        "d = \"KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP\"\n",
        "print(pairwise(c,d))"
      ],
      "metadata": {
        "id": "rGZOz_OOFlrk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Second Option\n",
        "def globalAlignment2():\n",
        "  M, Ix, Iy, traceM, traceIx, traceIy = createMatrices()\n",
        "\n",
        "  # Initializing the first column and first row\n",
        "  Ix[0][0] = Iy[0][0] = gop\n",
        "  for i in range(1, len(row)):\n",
        "    M[i][0] = Iy[i][0] = np.NINF\n",
        "    Ix[i][0] = gop + i * gex \n",
        "  for j in range(1, len(col)):\n",
        "    M[0][j] = Ix[0][j] = np.NINF\n",
        "    Iy[0][j] = gop + j * gex\n",
        "\n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      mat = score.loc[row[i]][col[j]]\n",
        "      M[i][j] = max((M[i-1][j-1] + mat), (Ix[i-1][j-1] + mat), (Iy[i-1][j-1] + mat)) \n",
        "      Ix[i,j] = max((M[i - 1, j] + gop), (Ix[i - 1, j] + gex))\n",
        "      Iy[i,j] = max((M[i, j - 1] + gop), (Iy[i, j - 1] + gex))\n",
        "\n",
        "      if M[i][j] == (M[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'm'\n",
        "      elif M[i][j] == (Ix[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'x'\n",
        "      elif M[i][j] == (Iy[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'y'\n",
        "\n",
        "      if Ix[i][j] == (M[i-1][j] + gop):\n",
        "        traceIx[i][j] = 'm'\n",
        "      else:\n",
        "        traceIx[i][j] = 'x'\n",
        "      \n",
        "      if Iy[i][j] == (M[i][j-1] + gop):\n",
        "        traceIy[i][j] = 'm'\n",
        "      else:\n",
        "        traceIy[i][j] = 'y'    \n",
        "\n",
        "# Trace back section for global alignment\n",
        "#def traceBack():\n",
        "  j = len(seq1)\n",
        "  i = len(seq2)\n",
        "  aln1 = ''\n",
        "  aln2 = ''\n",
        "  mxm = max(M[i][j], Ix[i][j], Iy[i][j])\n",
        "  tracer = ''\n",
        "  if mxm == M[i][j]:\n",
        "    tracer = 'm'\n",
        "  elif mxm == Ix[i][j]:\n",
        "    tracer = 'x'\n",
        "  else:\n",
        "    tracer = 'y'\n",
        "    \n",
        "  while (i > 0 and j > 0):\n",
        "    if tracer == 'm':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      #i, j = i - 1, j - 1\n",
        "      if traceM[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceM[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      elif traceM[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      i, j = i - 1, j - 1\n",
        "      continue\n",
        "    \n",
        "    elif tracer == 'x':\n",
        "      aln1 = '-' + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      #i = i - 1\n",
        "      if traceIx[i][j] == 'm':\n",
        "        tracer = 'm'    \n",
        "      elif traceIx[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      i = i - 1\n",
        "      continue\n",
        "      \n",
        "    elif tracer == 'y':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = '-' + aln2\n",
        "      if traceIy[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceIy[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      j = j - 1\n",
        "    else:\n",
        "      break\n",
        "      \n",
        "  return aln1, aln2\n",
        "# End of Trace Back for the Global Alignment\n",
        "# --------------------------------------------\n",
        "globalAlignment2()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "O-eBie3Sudo0",
        "outputId": "5c7303fc-dad3-4745-edb8-bc4e60af8243"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "('KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP',\n",
              " 'N-LFVALYDFVASGDNTLSITKGEKLRVLGYNHNG--EWCEAQTKNGQGWVPSNYITPVN')"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def localAlignment2():\n",
        "  bestScore = 0.0\n",
        "  # Filling score matrices and keeping track of the direction for each cell\n",
        "  M, Ix, Iy, traceM, traceIx, traceIy = createMatrices()\n",
        "\n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      mat = score.loc[row[i]][col[j]]\n",
        "      M[i][j] = max((M[i-1][j-1] + mat), (Ix[i-1][j-1] + mat), (Iy[i-1][j-1] + mat), 0.0) \n",
        "      Ix[i,j] = max((M[i - 1, j] + gop), (Ix[i - 1, j] + gex), 0.0)\n",
        "      Iy[i,j] = max((M[i, j - 1] + gop), (Iy[i, j - 1] + gex), 0.0)\n",
        "\n",
        "      if M[i][j] == (M[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'm'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "      elif M[i][j] == (Ix[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'x'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "      elif M[i][j] == (Iy[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'y'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "      \n",
        "\n",
        "      if Ix[i][j] == (Ix[i - 1, j] + gex):\n",
        "        traceIx[i][j] = 'x'\n",
        "        bestScore = isbest(Ix[i][j], bestScore)\n",
        "      elif Ix[i][j] == (M[i - 1, j] + gop):\n",
        "        traceIx[i][j] = 'm'\n",
        "        bestScore = isbest(Ix[i][j], bestScore)\n",
        "      \n",
        "      if Iy[i][j] == (Iy[i, j - 1] + gex):\n",
        "        traceIy[i][j] = 'y'\n",
        "        bestScore = isbest(Iy[i][j], bestScore)\n",
        "      elif Iy[i][j] == (M[i, j - 1] + gop):\n",
        "        traceIy[i][j] = 'm'   \n",
        "        bestScore = isbest(Iy[i][j], bestScore)\n",
        "\n",
        "# Trace back section for local alignment\n",
        "  r, c = M.shape # row and col lengths to track best score locations\n",
        "  bestLoc = [] # list of best score cells\n",
        "  pairs = [] # list of local pairs\n",
        "  \n",
        "  # Find all locations of the best score from the local alignment score matrix\n",
        "  for i in range(1, r):\n",
        "    for j in range(1, c):\n",
        "      if M[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "      if Ix[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "      if Iy[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "  \n",
        "  for k in range(len(bestLoc)): # loop over all optimal local alignments\n",
        "    i, j = bestLoc[k]\n",
        "    aln1 = ''\n",
        "    aln2 = ''\n",
        "    mxm = max(M[i][j], Ix[i][j], Iy[i][j], 0)\n",
        "    tracer = ''\n",
        "    if mxm == M[i][j]:\n",
        "      tracer = 'm'\n",
        "    elif mxm == Ix[i][j]:\n",
        "      tracer = 'x'\n",
        "    else:\n",
        "      tracer = 'y'\n",
        "      \n",
        "    while (i > 0 and j > 0):# and tracer != ''):\n",
        "      if tracer == 'm' and M[i][j]>0:\n",
        "        aln1 = col[j] + aln1\n",
        "        aln2 = row[i] + aln2\n",
        "        if traceM[i][j] == 'm':\n",
        "          tracer = 'm'\n",
        "        elif traceM[i][j] == 'x':\n",
        "          tracer = 'x'\n",
        "        elif traceM[i][j] == 'y':\n",
        "          tracer = 'y'\n",
        "        i, j = i - 1, j - 1\n",
        "        continue\n",
        "      \n",
        "      elif tracer == 'x' and Ix[i][j] > 0:\n",
        "        aln1 = '-' + aln1\n",
        "        aln2 = row[i] + aln2\n",
        "        if traceIx[i][j] == 'm':\n",
        "          tracer = 'm'    \n",
        "        elif traceIx[i][j] == 'x':\n",
        "          tracer = 'x'\n",
        "        i = i - 1\n",
        "        continue\n",
        "        \n",
        "      elif tracer == 'y' and Iy[i][j] > 0:\n",
        "        aln1 = col[j] + aln1\n",
        "        aln2 = '-' + aln2\n",
        "        if traceIy[i][j] == 'm':\n",
        "          tracer = 'm'\n",
        "        elif traceIy[i][j] == 'y':\n",
        "          tracer = 'y'\n",
        "        j = j - 1\n",
        "      else:\n",
        "        break\n",
        "    pairs.append((aln1, aln2))\n",
        "  return pairs  "
      ],
      "metadata": {
        "id": "-0epRHxO1T9w"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "def globalAlignment3():\n",
        "  M = np.zeros((len(row), len(col))) \n",
        "  Ix = np.zeros((len(row), len(col)))\n",
        "  Iy = np.zeros((len(row), len(col)))\n",
        "\n",
        "  traceM = np.zeros((len(row), len(col)), dtype = object)\n",
        "  traceIx = np.zeros((len(row), len(col)), dtype = object)\n",
        "  traceIy = np.zeros((len(row), len(col)), dtype = object)\n",
        "\n",
        "  Ix[0][0] = Iy[0][0] = gop\n",
        "  for i in range(1, len(row)):\n",
        "    M[i][0] = Iy[i][0] = np.NINF\n",
        "    Ix[i][0] = gop + i * gex \n",
        "  for j in range(1, len(col)):\n",
        "    M[0][j] = Ix[0][j] = np.NINF\n",
        "    Iy[0][j] = gop + j * gex\n",
        "\n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      mat = score.loc[row[i]][col[j]]\n",
        "      M[i][j] = max((M[i-1][j-1] + mat), (Ix[i-1][j-1] + mat), (Iy[i-1][j-1] + mat)) \n",
        "      Ix[i,j] = max((M[i - 1, j] + gop), (Ix[i - 1, j] + gex))\n",
        "      Iy[i,j] = max((M[i, j - 1] + gop), (Iy[i, j - 1] + gex))\n",
        "\n",
        "      if M[i][j] == (Ix[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'x'\n",
        "      elif M[i][j] == (Iy[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'y'\n",
        "      else:\n",
        "        traceM[i][j] = 'm'\n",
        "\n",
        "      if Ix[i][j] == (Ix[i - 1, j] + gex):\n",
        "        traceIx[i][j] = 'x'\n",
        "      else:\n",
        "        traceIx[i][j] = 'm'\n",
        "      \n",
        "      if Iy[i][j] == (Iy[i, j - 1] + gex):\n",
        "        traceIy[i][j] = 'y'\n",
        "      else:\n",
        "        traceIy[i][j] = 'm'   \n",
        "\n",
        "# Trace back  for global \n",
        "  j = len(seq1)\n",
        "  i = len(seq2)\n",
        "  aln1 = ''\n",
        "  aln2 = ''\n",
        "  mxm = max(M[i][j], Ix[i][j], Iy[i][j])\n",
        "  tracer = ''\n",
        "  if mxm == M[i][j]:\n",
        "    tracer = 'm'\n",
        "  elif mxm == Ix[i][j]:\n",
        "    tracer = 'x'\n",
        "  else:\n",
        "    tracer = 'y'\n",
        "    \n",
        "  while (i > 0 and j > 0):\n",
        "    if tracer == 'm':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      if traceM[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceM[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      elif traceM[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      i, j = i - 1, j - 1\n",
        "      continue\n",
        "    \n",
        "    elif tracer == 'x':\n",
        "      aln1 = '-' + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      if traceIx[i][j] == 'm':\n",
        "        tracer = 'm'    \n",
        "      elif traceIx[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      i = i - 1\n",
        "      continue\n",
        "      \n",
        "    elif tracer == 'y':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = '-' + aln2\n",
        "      if traceIy[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceIy[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      j = j - 1\n",
        "    else:\n",
        "      break\n",
        "      \n",
        "  return aln1, aln2\n",
        "\n",
        "def isbest(a, b):\n",
        "  if float(a) > float(b):\n",
        "    return a\n",
        "  else:\n",
        "    return b\n",
        "\n",
        "# Local Alignment  \n",
        "def localAlignment3():\n",
        "  bestScore = 0.0\n",
        "  # Filling score matrices and keeping track of the direction for each cell\n",
        "  M = np.zeros((len(row), len(col))) \n",
        "  Ix = np.zeros((len(row), len(col)))\n",
        "  Iy = np.zeros((len(row), len(col)))\n",
        "\n",
        "  traceM = np.zeros((len(row), len(col)), dtype = object)\n",
        "  traceIx = np.zeros((len(row), len(col)), dtype = object)\n",
        "  traceIy = np.zeros((len(row), len(col)), dtype = object)\n",
        "\n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      mat = score.loc[row[i]][col[j]]\n",
        "      M[i][j] = max((M[i-1][j-1] + mat), (Ix[i-1][j-1] + mat), (Iy[i-1][j-1] + mat), 0.0) \n",
        "      Ix[i,j] = max((M[i - 1, j] + gop), (Ix[i - 1, j] + gex), 0.0)\n",
        "      Iy[i,j] = max((M[i, j - 1] + gop), (Iy[i, j - 1] + gex), 0.0)\n",
        "\n",
        "      if M[i][j] == (Ix[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'x'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "      elif M[i][j] == (Iy[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'y'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "      elif M[i][j] == (M[i-1][j-1] + mat):\n",
        "        traceM[i][j] = 'm'\n",
        "        bestScore = isbest(M[i][j], bestScore)\n",
        "\n",
        "      if Ix[i][j] == (Ix[i - 1, j] + gex):\n",
        "        traceIx[i][j] = 'x'\n",
        "        bestScore = isbest(Ix[i][j], bestScore)\n",
        "      elif Ix[i][j] == (M[i - 1, j] + gop):\n",
        "        traceIx[i][j] = 'm'\n",
        "        bestScore = isbest(Ix[i][j], bestScore)\n",
        "      \n",
        "      if Iy[i][j] == (Iy[i, j - 1] + gex):\n",
        "        traceIy[i][j] = 'y'\n",
        "        bestScore = isbest(Iy[i][j], bestScore)\n",
        "      elif Iy[i][j] == (M[i, j - 1] + gop):\n",
        "        traceIy[i][j] = 'm'   \n",
        "        bestScore = isbest(Iy[i][j], bestScore)\n",
        "\n",
        "# Trace back for local \n",
        "  r, c = M.shape # row and col lengths to track best score locations\n",
        "  bestLoc = (0,0) \n",
        "  \n",
        "  # Find  location of the best score \n",
        "  for i in range(1, r):\n",
        "    for j in range(1, c):\n",
        "      if M[i][j] == bestScore:\n",
        "        bestLoc = (i, j)\n",
        "      if Ix[i][j] == bestScore:\n",
        "        bestLoc = (i, j)\n",
        "      if Iy[i][j] == bestScore:\n",
        "        bestLoc = (i, j)\n",
        "  \n",
        "  i, j = bestLoc\n",
        "  aln1 = ''\n",
        "  aln2 = ''\n",
        "  mxm = max(M[i][j], Ix[i][j], Iy[i][j], 0)\n",
        "  tracer = ''\n",
        "  if mxm == M[i][j]:\n",
        "    tracer = 'm'\n",
        "  elif mxm == Ix[i][j]:\n",
        "    tracer = 'x'\n",
        "  else:\n",
        "    tracer = 'y'\n",
        "      \n",
        "  while (i > 0 and j > 0):\n",
        "    if tracer == 'm' and M[i][j]>0:\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      if traceM[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceM[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      elif traceM[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      i, j = i - 1, j - 1\n",
        "      continue\n",
        "      \n",
        "    elif tracer == 'x' and Ix[i][j] > 0:\n",
        "      aln1 = '-' + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      if traceIx[i][j] == 'm':\n",
        "        tracer = 'm'    \n",
        "      elif traceIx[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      i = i - 1\n",
        "      continue\n",
        "        \n",
        "    elif tracer == 'y' and Iy[i][j] > 0:\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = '-' + aln2\n",
        "      if traceIy[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceIy[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      j = j - 1\n",
        "    else:\n",
        "      break\n",
        "  return aln1, aln2\n",
        "localAlignment3()"
      ],
      "metadata": {
        "id": "E_afhls_9xDm",
        "outputId": "7b5f29a4-5ddd-4c5b-f1d7-c818222bee2b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "('VIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLG',\n",
              " 'LFVALYDFVASGDNTLSITKGEKLRVLGYNHNGE--WCEAQTKNGQGWVPSNYIT')"
            ]
          },
          "metadata": {},
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "globalAlignment3()"
      ],
      "metadata": {
        "id": "5V11e6tGKXlp",
        "outputId": "3d6aeb7a-eae1-4313-a786-68a991c2f578",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "('KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP',\n",
              " 'N-LFVALYDFVASGDNTLSITKGEKLRVLGYNHNGE--WCEAQTKNGQGWVPSNYITPVN')"
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "cmwv0kp1RjtI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**HW5: UPGMA** "
      ],
      "metadata": {
        "id": "LHYwDzaGBhN8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!npx degit https://github.com/mequanent/Bioinformatics-hw2 -f"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OiONMNipPs98",
        "outputId": "fbb3f11a-df86-49d3-a8d2-566fdf7deb0a"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[K\u001b[?25hnpx: installed 1 in 3.452s\n",
            "\u001b[36m> destination directory is not empty. Using --force, continuing\u001b[39m\n",
            "\u001b[36m> cloned \u001b[1mmequanent/Bioinformatics-hw2\u001b[22m#\u001b[1mHEAD\u001b[22m\u001b[39m\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "pd.set_option('display.max_colwidth', None)\n",
        "\n",
        "input = \"distanceMatrix1.txt\"\n",
        "input = open(input, 'r')\n",
        "lines = input.readlines()[1:]\n",
        "input.close()\n",
        "\n",
        "distance = []\n",
        "indices = []\n",
        "i = 0\n",
        "for line in lines:\n",
        "  line = line.split()\n",
        "  distance.append(line)\n",
        "  indices.append(i)\n",
        "  i += 1\n",
        "size = len(indices)\n",
        "distance = np.array(distance).astype(float)\n",
        "dist = pd.DataFrame(distance, index = indices, columns = indices)\n",
        "#dist"
      ],
      "metadata": {
        "id": "6PUll19PBd_o"
      },
      "execution_count": 57,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def mini(a):\n",
        "  loc = (0, 0)\n",
        "  min = np.Inf\n",
        "  for i in range(1, a.shape[0]):\n",
        "    for j in range(i):\n",
        "      if a.loc[i][j] < min:\n",
        "        min = a.loc[i][j]\n",
        "        loc = (i, j)  \n",
        "  return min, loc\n",
        "\n",
        "mini(dist)\n",
        "dct = dict()\n",
        "distance[5][:5]"
      ],
      "metadata": {
        "id": "Dycxhlmgbwdc",
        "outputId": "7fb91fcf-3498-40a7-a905-f09de238e56b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 58,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([1.6043, 1.4389, 0.6179, 0.5061, 0.3484])"
            ]
          },
          "metadata": {},
          "execution_count": 58
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def UPGMA():\n",
        "  newdist = dist.copy()\n",
        "  # Add additional column to store the combined entities\n",
        "  newdist['paired'] = [str(i) for i in indices]\n",
        "  newdist['paired_no_dist'] = [str(i) for i in indices]\n",
        "\n",
        "  while len(newdist) > 1:\n",
        "    min_dist, loc = mini(newdist)\n",
        "    half = str(round(min_dist / 2, 4)) # round result to 4 decimal places\n",
        "\n",
        "    i, j = min(loc), max(loc)\n",
        "    p1, p2 = newdist.loc[i]['paired'], newdist.loc[j]['paired']\n",
        "    p11, p22 = newdist.loc[i]['paired_no_dist'], newdist.loc[j]['paired_no_dist']\n",
        "\n",
        "    dist1 = newdist.loc[i][~newdist.columns.isin(['paired', 'paired_no_dist'])] # ~df.columns.isin(\n",
        "    dist2 = newdist.loc[j][~newdist.columns.isin(['paired', 'paired_no_dist'])]\n",
        "    average = (dist1 + dist2)  / 2\n",
        "\n",
        "    pair = \"(\" + p2 + ':' + half + \", \" + p1 + ':' + half + \")\"  # form paired string\n",
        "    pair_no_dist = \"(\" + p22 +  \", \" + p11 + \")\" \n",
        "\n",
        "    newdist[i] = average # replace column with the average value\n",
        "    newdist.loc[i] = average # replace row with the average value\n",
        "\n",
        "    newdist.drop(j, inplace = True)\n",
        "    newdist.drop(j, inplace = True, axis = 1)\n",
        "    newdist.loc[i, i] = 0\n",
        "\n",
        "    newdist.loc[i, 'paired'] = pair\n",
        "    newdist.loc[i, 'paired_no_dist'] = pair_no_dist\n",
        "    #print(newdist)\n",
        "    #print('*************************************')\n",
        "  return newdist['paired'][0], newdist['paired_no_dist'][0]\n",
        "print(UPGMA()[0])\n",
        "print(UPGMA()[1])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "--NuerU_caFj",
        "outputId": "4e878c03-03cc-410b-c107-30afdaa923a7"
      },
      "execution_count": 91,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "((((((6:0.1346, 5:0.1346):0.1642, 4:0.1642):0.2379, 3:0.2379):0.3259, 2:0.3259):0.7474, 1:0.7474):0.838, 0:0.838)\n",
            "((((((6, 5), 4), 3), 2), 1), 0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "N8_MzP7hMdZ9"
      },
      "execution_count": 91,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Get off diagonal part\n",
        "def offdiagonal(a):\n",
        "  offdiag = []\n",
        "  for i in range(1, a.shape[0]):\n",
        "    row = []\n",
        "    for j in range(i):\n",
        "      row.append(a[i][j])\n",
        "    offdiag.append(row)\n",
        "  return offdiag\n",
        "offdiagonal(distance)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EFY4q_wws6LY",
        "outputId": "cb7081f1-e3f6-473e-c9eb-07795cb7fe9b"
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[[1.6866],\n",
              " [1.7198, 1.5232],\n",
              " [1.6606, 1.4841, 0.7115],\n",
              " [1.5243, 1.4465, 0.5958, 0.4631],\n",
              " [1.6043, 1.4389, 0.6179, 0.5061, 0.3484],\n",
              " [1.5905, 1.4629, 0.5583, 0.471, 0.3083, 0.2692]]"
            ]
          },
          "metadata": {},
          "execution_count": 34
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "QVANPt1vv5Cf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "Jk3APW5D9NyM"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}