{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "hw1_110761504.ipynb",
      "provenance": [],
      "authorship_tag": "ABX9TyP6uimZszwgCs3Txo/rkmoH",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mequanent/Bioinformatics-hw2/blob/main/hws_110761504.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Clone the repository first using the following command\n",
        "!npx degit https://github.com/mequanent/Bioinformatics-hw2 -f"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Eim-tkyVaZrg",
        "outputId": "b5553385-872a-451f-edaf-2ce96f9e10dc"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[K\u001b[?25hnpx: installed 1 in 1.998s\n",
            "\u001b[36m> destination directory is not empty. Using --force, continuing\u001b[39m\n",
            "\u001b[36m> cloned \u001b[1mmequanent/Bioinformatics-hw2\u001b[22m#\u001b[1mHEAD\u001b[22m\u001b[39m\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Homework 2**"
      ],
      "metadata": {
        "id": "GCOZucxBv9DA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "pip install Bio -q "
      ],
      "metadata": {
        "id": "Tba1tOZ5vhIb",
        "outputId": "f2c8e7ee-27ee-4544-b0f6-6cf7e368f811",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[K     |████████████████████████████████| 269 kB 12.8 MB/s \n",
            "\u001b[K     |████████████████████████████████| 2.3 MB 38.6 MB/s \n",
            "\u001b[?25h"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "#To install the Bio library into my IDLE python editor I used the following two steps.\n",
        "#    1. cd C:\\Users\\USER\\AppData\\Local\\Programs\\Python\\Python39\\Scripts\n",
        "#    2. pip install Bio\n",
        "\n",
        "# Read fasta and parse the seq_names and the sequences\n",
        "# Reference for parsing the fasta is the original biopython library documentation\n",
        "# (https://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc11 section 2.4.1)\n",
        "\n",
        "from Bio import SeqIO\n",
        "seq_name = []\n",
        "sequence = []\n",
        "# Remember to customize as --input from CLI args\n",
        "for seq_record in SeqIO.parse(\"test.fasta\", \"fasta\"):\n",
        "  seq_name.append(seq_record.id)\n",
        "  #sequence.append(repr(seq_record.seq))\n",
        "  sequence.append(seq_record.seq)\n",
        "  #print(len(seq_record))\n",
        "print(sequence[0])\n",
        "print(sequence[1])\n",
        "\n",
        "def preprocess_score(score):\n",
        "  file = open(score, 'r')\n",
        "  lines = file.readlines()\n",
        "  file.close()\n",
        "  columns = []\n",
        "  mut = []\n",
        "  for line in lines:\n",
        "    if line.startswith(\"#\") or line.startswith(\" \"):\n",
        "      continue\n",
        "    columns.append(line[0]) # Amino acid alphabets stored for row & column indexing\n",
        "    line = line[1:].strip().split() # remove the letters from left and the spaces\n",
        "    mut.append(line)\n",
        "  mut = np.array(mut).astype(int) # change the string numbers into type int\n",
        "  \n",
        "  return pd.DataFrame(mut, index = columns, columns = columns)\n",
        "\n",
        "score = \"pam250.txt\" # score from command line args\n",
        "score = preprocess_score(score) # score made a data frame to access its values with \n",
        "                  # alphabet indices\n",
        "#print(score)\n",
        "\n",
        "gop = -10\n",
        "gex = -2\n",
        "# A function to determine pairwise alignment score\n",
        "def pairwise(x, y):\n",
        "  pair_sum = 0\n",
        "  x_prev = '' # x_prev and y_prev are used to store immediate previous alphabets\n",
        "  y_prev = '' # so that extension gaps will be identified\n",
        "  for i in range(len(x)): # Considering that both x and y have the same length\n",
        "    x_prev = x_prev == '-' and x[i] == '-' # status of extension gap in the first paired sequence\n",
        "    y_prev = y_prev == '-' and y[i] == '-' # status of extension gap in second paired sequence\n",
        "    curr = x[i] == '-' or y[i] == '-' # status of open gap in the paired sequence\n",
        "    if x_prev or y_prev: # extension gap value will be taken if holds\n",
        "      pair_sum = pair_sum + gex\n",
        "    elif curr: # open gap value will be taken if holds\n",
        "      pair_sum = pair_sum + gop\n",
        "    else: # score matrix value will be taken otherwise\n",
        "      pair_sum = pair_sum + score.loc[x[i]][y[i]] \n",
        "    x_prev = x[i]\n",
        "    y_prev = y[i]\n",
        "  return pair_sum\n",
        "\n",
        "#pairwise(sequence[0], sequence[1])\n",
        "\n",
        "# A function to find the total sum of pairs score.\n",
        "def sumOfPairs(seq):\n",
        "  sop = 0 # to store sum of pairs value\n",
        "  for i in range(len(seq)-1): # i iterates as long as there is j in seq as its next sequence\n",
        "    for j in range(i+1, len(seq)): # j will be the next seq index to be compared with seq[i]\n",
        "      sop = sop + pairwise(seq[i], seq[j])\n",
        "  return sop\n",
        "print(sumOfPairs(sequence))"
      ],
      "metadata": {
        "id": "UlUbMGY6cAcx",
        "outputId": "5040bee1-cca7-41e6-8677-0046dd3049c7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "N-LFVALYDFVASGDNTLSITKGEKLRVLGYNH--NGEWCEAQTKNGQGWVPSNYITPVN\n",
            "KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP\n",
            "1047\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Homework 3**"
      ],
      "metadata": {
        "id": "whhnJvDTvyh4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "seq_name = []\n",
        "sequence = []\n",
        "# Remember to customize as --input from CLI args\n",
        "for seq_record in SeqIO.parse(\"test3.fasta\", \"fasta\"):\n",
        "  seq_name.append(seq_record.id)\n",
        "  sequence.append(seq_record.seq)\n",
        "\n",
        "seq1 = sequence[0]\n",
        "seq2 = sequence[1]\n",
        "\n",
        "print(\"+----- Given Sequences with their respective lengthes ---------+\")\n",
        "print(seq1, len(seq1))\n",
        "print(seq2, len(seq2))\n",
        "print(\"****************************************************************\")\n",
        "gap = -10\n",
        "score = \"pam250.txt\" # score from command line args\n",
        "aln = 'local' # Alignment choice argument from command line interface\n",
        "\n",
        "TOP = gap\n",
        "DIAGONAL = 0\n",
        "LEFT = gap\n",
        "col = '-' + seq1 # This if for simplicity to avoid repeated use of (len(seq1) + 1)\n",
        "row = '-' + seq2 # For the same reason\n",
        "\n",
        "def preprocess_score(score):\n",
        "  file = open(score, 'r')\n",
        "  lines = file.readlines()\n",
        "  file.close()\n",
        "  columns = []\n",
        "  mut = []\n",
        "  for line in lines:\n",
        "    if line.startswith(\"#\") or line.startswith(\" \"):\n",
        "      continue\n",
        "    columns.append(line[0]) # Amino acid alphabets stored for row & column indexing\n",
        "    line = line[1:].strip().split() # remove the letters from left and the spaces\n",
        "    mut.append(line)\n",
        "  mut = np.array(mut).astype(int) # change the string numbers into type int\n",
        "  return pd.DataFrame(mut, index = columns, columns = columns)\n",
        "\n",
        "score = preprocess_score(score) # score made a data frame to access its values  \n",
        "                  # with alphabet indices\n",
        "\n",
        "# *************************** Global Alignment Section *************************\n",
        "# Filling score_m and keeping track of the direction for each cell\n",
        "def globalAlignment(score):\n",
        "  score_m = np.zeros((len(row), len(col))) # score matrix for global alignment \n",
        "  direction_m = np.empty((len(row), len(col)), dtype = object)\n",
        "  for i in range(len(col)):\n",
        "    score_m[0][i] = i * gap\n",
        "  for j in range(len(row)):\n",
        "    score_m[j][0] = j * gap\n",
        "\n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      TOP = score_m[i-1,j] + gap\n",
        "      DIAGONAL = score_m[i-1][j-1] + score.loc[row[i]][col[j]]\n",
        "      LEFT = score_m[i,j-1] + gap\n",
        "      score_m[i][j] = max(TOP, DIAGONAL, LEFT) \n",
        "      # ------ Let the direction for each cell be saved as follows \n",
        "      if score_m[i][j] == TOP:\n",
        "        direction_m[i][j] = 'ins'\n",
        "      elif score_m[i][j] == DIAGONAL:\n",
        "        direction_m[i][j] = 'sub' \n",
        "      else:\n",
        "        direction_m[i][j] = 'del'\n",
        "\n",
        "  score_m = score_m.astype(int)\n",
        "  return score_m, direction_m\n",
        "\n",
        "# Trace back section for global alignment\n",
        "def traceBackGlobal():\n",
        "  score_m, direction_m = globalAlignment(score)\n",
        "  i = len(seq1)\n",
        "  j = len(seq2)\n",
        "  aln1 = ''\n",
        "  aln2 = ''\n",
        "\n",
        "  while(not (i == 0 and j == 0)):\n",
        "    if(direction_m[j][i] == 'sub'):\n",
        "      aln1 = col[i] + aln1\n",
        "      aln2 = row[j] + aln2\n",
        "      i, j = i - 1, j - 1\n",
        "    elif(direction_m[j][i] == 'del'):\n",
        "      aln1 = col[i] + aln1\n",
        "      aln2 = '-' + aln2\n",
        "      i = i - 1\n",
        "    else:\n",
        "      aln1 = '-' + aln1\n",
        "      aln2 = row[j] + aln2\n",
        "      j = j - 1\n",
        "  return aln1, aln2\n",
        "# End of Trace Back for the Global Alignment\n",
        "# --------------------------------------------\n",
        "# ************* End of Global Alignment ************************************\n",
        "\n",
        "def showMatched(x, y): # A function to help proper display of the alignments\n",
        "  between = ''\n",
        "  for i in range(len(x)):\n",
        "    if x[i] == y[i]:\n",
        "      between = between + '|'\n",
        "    elif x[i] == '-' or y[i] == '-':\n",
        "      between = between + ' '\n",
        "    else:\n",
        "      between = between + '.'\n",
        "  return between\n",
        "\n",
        "def displayAlignment(x, y): # A function to display the aligned sequences with their SoP score\n",
        "  print(\"+----- You wanted {} alignment and here is the result -----+\".format(aln))\n",
        "  print(x)\n",
        "  print(showMatched(x, y))\n",
        "  print(y)\n",
        "  print(\"Sum of pairs (SoP) score for the {a} alignment is: {b}\".format(a = aln, b = pairwise(x, y)))\n",
        "# The SoP value verifies the correctness of our alignment\n",
        "                # if it becomes equal to the right bottom value.\n",
        "\n",
        "# ************* Local Alignment ********************************************\n",
        "def localAlignment(score):\n",
        "  score_local = np.zeros((len(row), len(col))).astype(int) # score matrix for local alignment\n",
        "  direction_l = np.zeros((len(row), len(col)), dtype = object) # to trace direction\n",
        "  bestScore = 0\n",
        "  \n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      TOP = score_local[i-1,j] + gap\n",
        "      DIAGONAL = score_local[i-1][j-1] + score.loc[row[i]][col[j]]\n",
        "      LEFT = score_local[i,j-1] + gap\n",
        "      score_local[i][j] = max(TOP, DIAGONAL, LEFT, 0)\n",
        "\n",
        "      if score_local[i][j] >= bestScore:\n",
        "        bestScore = score_local[i][j]\n",
        "        \n",
        "      if score_local[i][j] == DIAGONAL:\n",
        "        direction_l[i][j] = 'sub'\n",
        "      elif score_local[i][j] == TOP:\n",
        "        direction_l[i][j] = 'ins' \n",
        "      else:\n",
        "        direction_l[i][j] = 'del'\n",
        "  return score_local, direction_l, bestScore\n",
        "\n",
        "# Trace back all possible optimal local alignment\n",
        "def traceBackLocal():\n",
        "  score_local, direction_l, bestScore = localAlignment(score)\n",
        "\n",
        "  r, c = score_local.shape # row and col lengths to track best score locations\n",
        "  bestLoc = [] # list of best score cells\n",
        "  pairs = [] # list of local pairs\n",
        "  \n",
        "  # Find all locations of the best score from the local alignment score matrix\n",
        "  for i in range(1, r):\n",
        "    for j in range(1, c):\n",
        "      if score_local[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "  \n",
        "  for k in range(len(bestLoc)): # loop over all optimal local alignments\n",
        "    i, j = bestLoc[k]\n",
        "    al1 = ''\n",
        "    al2 = ''\n",
        "    between = ''\n",
        "    while(score_local[i][j]> 0):\n",
        "      if(direction_l[i][j] == 'sub'):\n",
        "        al1 = col[j] + al1\n",
        "        al2 = row[i] + al2\n",
        "        i, j = i - 1, j - 1\n",
        "      elif(direction_l[i][j] == 'del'):\n",
        "        al1 = col[j] + al1\n",
        "        al2 = '-' + al2\n",
        "        j = j - 1\n",
        "      else:\n",
        "        al1 = '-' + al1\n",
        "        al2 = row[i] + al2\n",
        "        i = i - 1\n",
        "    pairs.append((al1, al2))\n",
        "  return pairs\n",
        "\n",
        "alignments = []\n",
        "if aln == 'global':\n",
        "  aln1, aln2 = traceBackGlobal()\n",
        "  displayAlignment(aln1, aln2)\n",
        "  alignments.append(SeqRecord(Seq(aln1), id=\"protien 1\"))\n",
        "  alignments.append(SeqRecord(Seq(aln2), id=\"protien 2\"))\n",
        "elif aln == 'local':\n",
        "  for i in traceBackLocal():\n",
        "    displayAlignment(i[0], i[1])\n",
        "    alignments.append(SeqRecord(Seq(i[0]), id=\"protien 1\"))\n",
        "    alignments.append(SeqRecord(Seq(i[1]), id=\"protien 2\"))\n",
        "else:\n",
        "  print(\"Please selected the alignment type.\")\n",
        "\n",
        "SeqIO.write(alignments, \"result.fasta\", \"fasta\")"
      ],
      "metadata": {
        "id": "e_YVQGnpvxOa",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "37ea9489-5b8f-4c75-853b-feae5b4c6b59"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "+----- Given Sequences with their respective lengthes ---------+\n",
            "NLFVALYDFVASGDNTLSITKGEKLRVLGYNHNGEWCEAQTKNGQGWVPSNYITPVN 57\n",
            "KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP 60\n",
            "****************************************************************\n",
            "+----- You wanted local alignment and here is the result -----+\n",
            "LFVALYDFVASGDNTLSITKGEKLRVLG-YNHNG-EWCEAQTKNGQGWVPSNYI\n",
            "...||.|.....|..|....|....... ..... ||..|......|.||.|..\n",
            "VIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLL\n",
            "Sum of pairs (SoP) score for the local alignment is: 52\n",
            "+----- You wanted local alignment and here is the result -----+\n",
            "LFVALYDFVASGDNTLSITKGEKLRVLG-YNHNG-EWCEAQTKNGQGWVPSNYIT\n",
            "...||.|.....|..|....|....... ..... ||..|......|.||.|...\n",
            "VIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLG\n",
            "Sum of pairs (SoP) score for the local alignment is: 52\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "4"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**H4 pre**"
      ],
      "metadata": {
        "id": "z7C25q8Op5_o"
      }
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "y1n75yCW5qmz",
        "outputId": "33102500-b899-448e-c925-c0f2bfc84b6d",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "('N-LFVALYDFVASGDNTLSITKGEKLRVLGYNHNGE--WCEAQTKNGQGWVPSNYITPVN',\n",
              " 'KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP')"
            ]
          },
          "metadata": {},
          "execution_count": 40
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Second Option\n",
        "M = np.zeros((len(row), len(col))) # score matrix for global alignment \n",
        "Ix = np.zeros((len(row), len(col)))\n",
        "Iy = np.zeros((len(row), len(col)))\n",
        "\n",
        "traceM = np.zeros((len(row), len(col)), dtype = object)\n",
        "traceIx = np.zeros((len(row), len(col)), dtype = object)\n",
        "traceIy = np.zeros((len(row), len(col)), dtype = object)\n",
        "\n",
        "# Initializing the first column and first row\n",
        "Ix[0][0] = Iy[0][0] = gop\n",
        "for i in range(1, len(row)):\n",
        "  M[i][0] = Iy[i][0] = np.NINF\n",
        "  Ix[i][0] = gop + i * gex \n",
        "for j in range(1, len(col)):\n",
        "  M[0][j] = Ix[0][j] = np.NINF\n",
        "  Iy[0][j] = gop + j * gex\n",
        "Ix.shape\n",
        "\n",
        "for i in range(1, len(row)):\n",
        "  for j in range(1, len(col)):\n",
        "    mat = score.loc[row[i]][col[j]]\n",
        "    M[i][j] = max((M[i-1][j-1] + mat), (Ix[i-1][j-1] + mat), (Iy[i-1][j-1] + mat)) \n",
        "    Ix[i,j] = max((M[i - 1, j] + gop), (Ix[i - 1, j] + gex))\n",
        "    Iy[i,j] = max((M[i, j - 1] + gop), (Iy[i, j - 1] + gex))\n",
        "\n",
        "    if M[i][j] == (M[i-1][j-1] + mat):\n",
        "      traceM[i][j] = 'm'\n",
        "    elif M[i][j] == (Ix[i-1][j-1] + mat):\n",
        "      traceM[i][j] = 'x'\n",
        "    elif M[i][j] == (Iy[i-1][j-1] + mat):\n",
        "      traceM[i][j] = 'y'\n",
        "\n",
        "    if Ix[i][j] == (M[i-1][j] + gop):\n",
        "      traceIx[i][j] = 'm'\n",
        "    else:\n",
        "      traceIx[i][j] = 'x'\n",
        "    \n",
        "    if Iy[i][j] == (M[i][j-1] + gop):\n",
        "      traceIy[i][j] = 'm'\n",
        "    else:\n",
        "      traceIy[i][j] = 'y'    \n",
        "\n",
        "# Trace back section for global alignment\n",
        "def traceBack():\n",
        "  j = len(seq1)\n",
        "  i = len(seq2)\n",
        "  aln1 = ''\n",
        "  aln2 = ''\n",
        "  mxm = max(M[i][j], Ix[i][j], Iy[i][j])\n",
        "  tracer = ''\n",
        "  if mxm == M[i][j]:\n",
        "    tracer = 'm'\n",
        "  elif mxm == Ix[i][j]:\n",
        "    tracer = 'x'\n",
        "  else:\n",
        "    tracer = 'y'\n",
        "    \n",
        "  while (i > 0 and j > 0):\n",
        "    if tracer == 'm':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      #i, j = i - 1, j - 1\n",
        "      if traceM[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceM[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      elif traceM[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      i, j = i - 1, j - 1\n",
        "      continue\n",
        "    \n",
        "    elif tracer == 'x':\n",
        "      aln1 = '-' + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      #i = i - 1\n",
        "      if traceIx[i][j] == 'm':\n",
        "        tracer = 'm'    \n",
        "      elif traceIx[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      i = i - 1\n",
        "      continue\n",
        "      \n",
        "    elif traceIy[i][j] == 'y':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = '-' + aln2\n",
        "      if traceIy[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceIy[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      j = j - 1\n",
        "      \n",
        "  return aln1, aln2\n",
        "# End of Trace Back for the Global Alignment\n",
        "# --------------------------------------------\n",
        "traceBack()"
      ],
      "metadata": {
        "id": "j1Ef5PD6xmAe"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Expected Output \n",
        "c = \"N-LFVALYDFVASGDNTLSITKGEKLRVLGYNHNGE--WCEAQTKNGQGWVPSNYITPVN\"\n",
        "d = \"KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP\"\n",
        "print(pairwise(c,d))\n",
        "# Another option\n",
        "c = \"N-LFVALYDFVASGDNTLSITKGEKLRVLGYNHNG--EWCEAQTKNGQGWVPSNYITPVN\"\n",
        "d = \"KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP\"\n",
        "print(pairwise(c,d))"
      ],
      "metadata": {
        "id": "cdrBTJk44z9D",
        "outputId": "1b4ca221-b02b-4996-80f1-4226cc81c824",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "45\n",
            "45\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "a, b = traceBack()\n",
        "print(a[0:4])\n",
        "print(b[0:4])\n",
        "\n",
        "for i in range(5):\n",
        "  l = []\n",
        "  for j in range(4):\n",
        "    l.append((M[i][j], Ix[i][j], Iy[i][j]))\n",
        "  print(l)\n",
        "for i in range(5):\n",
        "  l = []\n",
        "  for j in range(4):\n",
        "    l.append((traceM[i][j], traceIx[i][j], traceIy[i][j]))\n",
        "  print(l)"
      ],
      "metadata": {
        "id": "ZeaTlAKqVbRw",
        "outputId": "60632445-7428-4232-870e-15c64463b93e",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "NL-F\n",
            "KGVI\n",
            "[(0.0, -10.0, -10.0), (-inf, -inf, -12.0), (-inf, -inf, -14.0), (-inf, -inf, -16.0)]\n",
            "[(-inf, -12.0, -inf), (1.0, -inf, -inf), (-15.0, -inf, -9.0), (-19.0, -inf, -11.0)]\n",
            "[(-inf, -14.0, -inf), (-12.0, -9.0, -inf), (-3.0, -25.0, -22.0), (-14.0, -29.0, -13.0)]\n",
            "[(-inf, -16.0, -inf), (-16.0, -11.0, -inf), (-7.0, -13.0, -26.0), (-4.0, -24.0, -17.0)]\n",
            "[(-inf, -18.0, -inf), (-18.0, -13.0, -inf), (-9.0, -15.0, -28.0), (-6.0, -14.0, -19.0)]\n",
            "[(0, 0, 0), (0, 0, 0), (0, 0, 0), (0, 0, 0)]\n",
            "[(0, 0, 0), ('m', 'm', 'm'), ('y', 'm', 'm'), ('y', 'm', 'y')]\n",
            "[(0, 0, 0), ('x', 'm', 'm'), ('m', 'm', 'm'), ('y', 'm', 'm')]\n",
            "[(0, 0, 0), ('x', 'x', 'm'), ('x', 'm', 'm'), ('m', 'm', 'm')]\n",
            "[(0, 0, 0), ('x', 'x', 'm'), ('x', 'x', 'm'), ('m', 'm', 'm')]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "score.loc['N']['K']\n",
        "pairwise(a,b)"
      ],
      "metadata": {
        "id": "W-LoNKUnSpRq",
        "outputId": "9a0769bd-9060-4fdb-b1c0-749a077968ec",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "39"
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#np.set_printoptions(threshold=np.nan)\n",
        "np.set_printoptions(threshold=np.inf, linewidth=np.nan)\n",
        "print(M)\n",
        "print(\"*Ix*\" * 20)\n",
        "print(Ix)\n",
        "print(\"*Iy*\" * 20)\n",
        "print(Iy)\n",
        "print(\"*TM*\" * 20)\n",
        "print(traceM)\n",
        "print(\"*TIx*\" * 20)\n",
        "print(traceIx)\n",
        "print(\"TIy*\" * 50)\n",
        "print(traceIy)"
      ],
      "metadata": {
        "id": "XTrsCvjjU3dW"
      },
      "execution_count": 41,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "***HW4***"
      ],
      "metadata": {
        "id": "FqM152ET55WU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "seq_name = []\n",
        "sequence = []\n",
        "# Remember to customize as --input from CLI args\n",
        "for seq_record in SeqIO.parse(\"test3.fasta\", \"fasta\"):\n",
        "  seq_name.append(seq_record.id)\n",
        "  sequence.append(seq_record.seq)\n",
        "\n",
        "seq1 = sequence[0]\n",
        "seq2 = sequence[1]\n",
        "\n",
        "print(\"+----- Given Sequences with their respective lengthes ---------+\")\n",
        "print(seq1, len(seq1))\n",
        "print(seq2, len(seq2))\n",
        "print(\"****************************************************************\")\n",
        "gop= -10\n",
        "gex = -2\n",
        "score = \"pam250.txt\" # score from command line args\n",
        "aln = 'global' # Alignment choice argument from command line interface\n",
        "\n",
        "col = '-' + seq1 # This if for simplicity to avoid repeated use of (len(seq1) + 1)\n",
        "row = '-' + seq2 # For the same reason\n",
        "\n",
        "def preprocess_score(score):\n",
        "  file = open(score, 'r')\n",
        "  lines = file.readlines()\n",
        "  file.close()\n",
        "  columns = []\n",
        "  mut = []\n",
        "  for line in lines:\n",
        "    if line.startswith(\"#\") or line.startswith(\" \"):\n",
        "      continue\n",
        "    columns.append(line[0]) # Amino acid alphabets stored for row & column indexing\n",
        "    line = line[1:].strip().split() # remove the letters from left and the spaces\n",
        "    mut.append(line)\n",
        "  mut = np.array(mut).astype(int) # change the string numbers into type int\n",
        "  return pd.DataFrame(mut, index = columns, columns = columns)\n",
        "\n",
        "score = preprocess_score(score) # score made a data frame to access its values  \n",
        "                  # with alphabet indices\n",
        "\n",
        "# *************************** Global Alignment Section *************************\n",
        "# Filling score matrices and keeping track of the direction for each cell\n",
        "M = np.zeros((len(row), len(col))).astype(int) # score matrix for global alignment \n",
        "Ix = np.zeros((len(row), len(col))).astype(int)\n",
        "Iy = np.zeros((len(row), len(col))).astype(int)\n",
        "\n",
        "traceM = np.zeros((len(row), len(col)), dtype = object)\n",
        "traceIx = np.zeros((len(row), len(col)), dtype = object)\n",
        "traceIy = np.zeros((len(row), len(col)), dtype = object)\n",
        "\n",
        "# Initializing the first column and first row\n",
        "Ix[0][0] = Iy[0][0] = gop\n",
        "for i in range(1, len(row)):\n",
        "  M[i][0] = Iy[i][0] = np.NINF\n",
        "  Ix[i][0] = gop + i * gex \n",
        "for j in range(1, len(col)):\n",
        "  M[0][j] = Ix[0][j] = np.NINF\n",
        "  Iy[0][j] = gop + j * gex\n",
        "Ix.shape\n",
        "\n",
        "for i in range(1, len(row)):\n",
        "  for j in range(1, len(col)):\n",
        "    mat = score.loc[row[i]][col[j]]\n",
        "    M[i][j] = max((M[i-1][j-1] + mat), (Ix[i-1][j-1] + mat), (Iy[i-1][j-1] + mat)) \n",
        "    Ix[i,j] = max((M[i - 1, j] + gop), (Ix[i - 1, j] + gex))\n",
        "    Iy[i,j] = max((M[i, j - 1] + gop), (Iy[i, j - 1] + gex))\n",
        "\n",
        "    if M[i][j] == (Ix[i-1][j-1] + mat):\n",
        "      traceM[i][j] = 'x'\n",
        "    elif M[i][j] == (Iy[i-1][j-1] + mat):\n",
        "      traceM[i][j] = 'y'\n",
        "    else:\n",
        "      traceM[i][j] = 'm'\n",
        "\n",
        "    if Ix[i][j] == (Ix[i - 1, j] + gex):\n",
        "      traceIx[i][j] = 'x'\n",
        "    else:\n",
        "      traceIx[i][j] = 'm'\n",
        "    \n",
        "    if Iy[i][j] == (Iy[i, j - 1] + gex):\n",
        "      traceIy[i][j] = 'y'\n",
        "    else:\n",
        "      traceIy[i][j] = 'm'   \n",
        "\n",
        "# Trace back section for global alignment\n",
        "def traceBack():\n",
        "  j = len(seq1)\n",
        "  i = len(seq2)\n",
        "  aln1 = ''\n",
        "  aln2 = ''\n",
        "  mxm = max(M[i][j], Ix[i][j], Iy[i][j])\n",
        "  tracer = ''\n",
        "  if mxm == M[i][j]:\n",
        "    tracer = 'm'\n",
        "  elif mxm == Ix[i][j]:\n",
        "    tracer = 'x'\n",
        "  else:\n",
        "    tracer = 'y'\n",
        "    \n",
        "  while (i > 0 and j > 0):\n",
        "    if tracer == 'm':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      #i, j = i - 1, j - 1\n",
        "      if traceM[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceM[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      elif traceM[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      i, j = i - 1, j - 1\n",
        "      continue\n",
        "    \n",
        "    elif tracer == 'x':\n",
        "      aln1 = '-' + aln1\n",
        "      aln2 = row[i] + aln2\n",
        "      #i = i - 1\n",
        "      if traceIx[i][j] == 'm':\n",
        "        tracer = 'm'    \n",
        "      elif traceIx[i][j] == 'x':\n",
        "        tracer = 'x'\n",
        "      i = i - 1\n",
        "      continue\n",
        "      \n",
        "    elif traceIy[i][j] == 'y':\n",
        "      aln1 = col[j] + aln1\n",
        "      aln2 = '-' + aln2\n",
        "      #j = j - 1\n",
        "      if traceIy[i][j] == 'm':\n",
        "        tracer = 'm'\n",
        "      elif traceIy[i][j] == 'y':\n",
        "        tracer = 'y'\n",
        "      j = j - 1\n",
        "      \n",
        "  return aln1, aln2\n",
        "# End of Trace Back for the Global Alignment\n",
        "# --------------------------------------------\n",
        "# ************* End of Global Alignment ************************************\n",
        "\n",
        "def showMatched(x, y): # A function to help proper display of the alignments\n",
        "  between = ''\n",
        "  for i in range(len(x)):\n",
        "    if x[i] == y[i]:\n",
        "      between = between + '|'\n",
        "    elif x[i] == '-' or y[i] == '-':\n",
        "      between = between + ' '\n",
        "    else:\n",
        "      between = between + '.'\n",
        "  return between\n",
        "\n",
        "def displayAlignment(x, y): # A function to display the aligned sequences with their SoP score\n",
        "  print(\"+----- You wanted {} alignment and here is the result -----+\".format(aln))\n",
        "  print(x)\n",
        "  print(showMatched(x, y))\n",
        "  print(y)\n",
        "  print(\"Sum of pairs (SoP) score for the {a} alignment is: {b}\".format(a = aln, b = pairwise(x, y)))\n",
        "# The SoP value verifies the correctness of our alignment\n",
        "                # if it becomes equal to the right bottom value.\n",
        "\n",
        "# ************* Local Alignment ********************************************\n",
        "def localAlignment(score):\n",
        "  score_local = np.zeros((len(row), len(col))).astype(int) # score matrix for local alignment\n",
        "  direction_l = np.zeros((len(row), len(col)), dtype = object) # to trace direction\n",
        "  bestScore = 0\n",
        "  \n",
        "  for i in range(1, len(row)):\n",
        "    for j in range(1, len(col)):\n",
        "      TOP = score_local[i-1,j] + gap\n",
        "      DIAGONAL = score_local[i-1][j-1] + score.loc[row[i]][col[j]]\n",
        "      LEFT = score_local[i,j-1] + gap\n",
        "      score_local[i][j] = max(TOP, DIAGONAL, LEFT, 0)\n",
        "\n",
        "      if score_local[i][j] >= bestScore:\n",
        "        bestScore = score_local[i][j]\n",
        "        \n",
        "      if score_local[i][j] == DIAGONAL:\n",
        "        direction_l[i][j] = 'sub'\n",
        "      elif score_local[i][j] == TOP:\n",
        "        direction_l[i][j] = 'ins' \n",
        "      else:\n",
        "        direction_l[i][j] = 'del'\n",
        "  return score_local, direction_l, bestScore\n",
        "\n",
        "# Trace back all possible optimal local alignment\n",
        "def traceBackLocal():\n",
        "  score_local, direction_l, bestScore = localAlignment(score)\n",
        "\n",
        "  r, c = score_local.shape # row and col lengths to track best score locations\n",
        "  bestLoc = [] # list of best score cells\n",
        "  pairs = [] # list of local pairs\n",
        "  \n",
        "  # Find all locations of the best score from the local alignment score matrix\n",
        "  for i in range(1, r):\n",
        "    for j in range(1, c):\n",
        "      if score_local[i][j] == bestScore:\n",
        "        bestLoc.append((i, j))\n",
        "  \n",
        "  for k in range(len(bestLoc)): # loop over all optimal local alignments\n",
        "    i, j = bestLoc[k]\n",
        "    al1 = ''\n",
        "    al2 = ''\n",
        "    between = ''\n",
        "    while(score_local[i][j]> 0):\n",
        "      if(direction_l[i][j] == 'sub'):\n",
        "        al1 = col[j] + al1\n",
        "        al2 = row[i] + al2\n",
        "        i, j = i - 1, j - 1\n",
        "      elif(direction_l[i][j] == 'del'):\n",
        "        al1 = col[j] + al1\n",
        "        al2 = '-' + al2\n",
        "        j = j - 1\n",
        "      else:\n",
        "        al1 = '-' + al1\n",
        "        al2 = row[i] + al2\n",
        "        i = i - 1\n",
        "    pairs.append((al1, al2))\n",
        "  return pairs\n",
        "\n",
        "alignments = []\n",
        "if aln == 'global':\n",
        "  aln1, aln2 = traceBack()\n",
        "  displayAlignment(aln1, aln2)\n",
        "  alignments.append(SeqRecord(Seq(aln1), id=\"protien 1\"))\n",
        "  alignments.append(SeqRecord(Seq(aln2), id=\"protien 2\"))\n",
        "elif aln == 'local':\n",
        "  for i in traceBackLocal():\n",
        "    displayAlignment(i[0], i[1])\n",
        "    alignments.append(SeqRecord(Seq(i[0]), id=\"protien 1\"))\n",
        "    alignments.append(SeqRecord(Seq(i[1]), id=\"protien 2\"))\n",
        "else:\n",
        "  print(\"Please selected the alignment type.\")\n",
        "\n",
        "SeqIO.write(alignments, \"result.fasta\", \"fasta\")"
      ],
      "metadata": {
        "id": "TxfYaYuqUvx7",
        "outputId": "db2a7d0b-3771-4c32-bd2e-1bef2c10b1e9",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 43,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "+----- Given Sequences with their respective lengthes ---------+\n",
            "NLFVALYDFVASGDNTLSITKGEKLRVLGYNHNGEWCEAQTKNGQGWVPSNYITPVN 57\n",
            "KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP 60\n",
            "****************************************************************\n",
            "+----- You wanted global alignment and here is the result -----+\n",
            "N-LFVALYDFVASGDNTLSITKGEKLRVLGYNHNGE--WCEAQTKNGQGWVPSNYITPVN\n",
            ". ...||.|.....|..|....|............|  |..|......|.||.|......\n",
            "KGVIYALWDYEPQNDDELPMKEGDCMTIIHREDEDEIEWWWARLNDKEGYVPRNLLGLYP\n",
            "Sum of pairs (SoP) score for the global alignment is: 45\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "2"
            ]
          },
          "metadata": {},
          "execution_count": 43
        }
      ]
    }
  ]
}